Allow renaming an environment group

anukul opened the pull request:
## Pull request type

<!-- Please try to limit your pull request to one type, submit multiple pull requests if needed. -->

Please check the type of change your PR introduces:

- [ ] Bugfix
- [x] Feature
- [ ] Other (please describe):

## Pull request checklist

Please check if your PR fulfills the following requirements:

- [ ] If it's a backend change, tests for the changes have been added and `go test ./...` runs successfully from the root folder
- [ ] If it's a frontend change, Prettier has been run
- [ ] Docs have been reviewed and added / updated if needed

Draft PR, can add tests / docs if this makes sense.

## What is the current behavior?

Issue Number: Closes #841 

## What is the new behavior?


https://user-images.githubusercontent.com/44864521/125695461-2feb9fc0-8656-48d9-91bd-e1add1fab868.mov



## Technical Spec/Implementation Notes

- Implemented a rough design for draft PR, taken from the Delete option.  
Success (close modal) and Failure (do nothing) behaviour is also similar.

anukul commented:
cc @abelanger5

abelanger5 commented:
Looks really promising @anukul! My apologies for not reviewing the draft PR earlier. A few notes:
1. Could you outline how you intend to avoid data loss if the server exits while serving the request? It seems that the way this is currently implemented, the configmap + secret are deleted before they are created again. 
2. Have you tested that this properly mounts secret env variables into a container? It might be helpful to document how the  secret data is read/mounted into the container (perhaps we should create a doc in `/docs/developing` with a description of this architecture). Let's say I have a "secret environment variable" named `TEST` in an environment group named `testing`. The ConfigMap will have the following data:

```yaml
metadata:
  name: testing
...
data:
  TEST: PORTERSECRET_testing
```

While the secret will look like:

```yaml
metadata:
  name: testing
...
data:
  TEST: <secret-data>
```

So the way that this gets mounted into the container is that the Helm chart [parses](https://github.com/porter-dev/porter-charts/blob/79f87670d8d836a90424129e10e350102b80f9ac/applications/web/templates/deployment.yaml#L92) the `PORTERSECRET` configmap values and references them as a secret with the following spec:

```yaml
... # pod spec
env:
  name: TEST
  valueFrom:
    secretKeyRef:
      name: testing
      key: TEST
```

All that said, the implementation looks like it will work with secret env variables, since the line `configMap.EnvVariables[key] = fmt.Sprintf("PORTERSECRET_%s", configMap.Name)` will overwrite the keys that are set to the "incorrect" secret value. But I'm not sure if there's a cleaner way to override these values.

3. It looks like the CRLF endings got replaced with LF endings in the file `/internal/kubernetes/agent.go`, creating a really large diff. Could you revert this? It makes the code a bit harder to review. While we should be using LF endings, this should be in a different PR. 

Once you'd like me to test it and review the code, just mark the PR as "ready to review". Thanks!

abelanger5 commented:
Hi @anukul, just checking in here! Let me know if there's anything else you need to complete this PR 👍

anukul commented:
@abelanger5 wrt the notes,

1. I have now moved the delete operation after the create operation. It's not "atomic", but no data loss as such. Is this acceptable behavior?
2. My local setup is a bit hacky so I haven't configured it to the point where I can actually run apps yet.
Either way, I don't think this PR involves overriding any values since it deletes the old secret and creates a new one, with the new name. I did test this part, so I believe it should work correctly since I haven't modified anything else.
3. Sure, I see it has now been updated, so I've rebased on master.

abelanger5 commented on server/api/k8s_handler.go:
```diff
@@ -559,6 +565,92 @@ func (app *App) HandleUpdateConfigMap(w http.ResponseWriter, r *http.Request) {
+		SecretEnvVariables: decodedSecretData,
+	}
+
+	if err := createConfigMap(agent, newConfigMap); err != nil {
```
Before this function call, we should check if the new name of the config map is different from the old name, otherwise the backend will throw a `500` error with the following error log:

```
{"level":"warn","error":"secrets \"test-env-rename-2\" already exists","errExt":"{\"code\":500,\"errors\":[\"internal server error\"]}","time":"2021-07-29T18:52:22Z"}
```

(ideally this is checked from the frontend as well)

anukul replied:
Oh right.. thanks, fixed!

abelanger5 commented on dashboard/src/main/home/cluster-dashboard/env-groups/ExpandedEnvGroup.tsx:
```diff
@@ -86,6 +91,30 @@ class ExpandedEnvGroup extends Component<PropsType, StateType> {
+        }
+      )
+      .then((res) => {
+        this.props.closeExpanded();
```
Instead of closing the expanded env group here, it seems more intuitive to just rename the expanded env group and re-pull the configmap data. What do you think?

anukul replied:
I agree, that makes more sense.
I've pushed a commit for this, it involved some refactoring to store env group in state.

abelanger5 replied:
The state refactor is cleaner as well, thanks!

abelanger5 reviewed (changes requested):
Hey @anukul, just two minor comments, but otherwise it's working really well! 

> I have now moved the delete operation after the create operation. It's not "atomic", but no data loss as such. Is this acceptable behavior?

Yes, this seems acceptable, and I'm not sure how we would make this an atomic...

anukul commented on dashboard/src/main/home/cluster-dashboard/env-groups/ExpandedEnvGroup.tsx:
```diff
@@ -49,31 +55,48 @@ class ExpandedEnvGroup extends Component<PropsType, StateType> {
-    const envGroupName = this.props.envGroup.metadata.name;
+  populateEnvGroup = (envGroup: any) => {
+    const {
+      metadata: { name, creationTimestamp: timestamp },
```
Side note - the `creationTimestamp` field is used to display "Last updated ...", which does not change when the user updates config map variables.

abelanger5 replied:
Ah good catch -- I've filed an issue (#946) which we can tackle in a later PR.

abelanger5 reviewed (approved):
Thanks so much @anukul! Approved and merged into `master`, should be on the hosted version later today 👍
